feat(OptimizelyConfig): Add new fields to OptimizelyConfig

yasirfolio3 opened the pull request:
## Summary
The following new public properties are added to OptimizelyConfig:
- sdkKey
- environmentKey
- attributes
- audiences
- events
- experimentRules and deliveryRules to OptimizelyFeature
- audiences to OptimizelyExperiment

## Test plan
All FSC tests OPTIMIZELY_CONFIG_V2 tests should pass.
FSC OptimizelyConfig [link](https://app.travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/235034135)

keppel2 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -37,20 +45,31 @@ export class OptimizelyConfig {
   constructor(configObj: ProjectConfig, datafile: string) {
-    this.experimentsMap = OptimizelyConfig.getExperimentsMap(configObj);
-    this.featuresMap = OptimizelyConfig.getFeaturesMap(configObj, this.experimentsMap);
+    this.sdkKey = configObj.sdkKey ? configObj.sdkKey : '';
```
You can use `configObj.sdkKey ?? '';` if you want.

zashraf1985 replied:
```suggestion
    this.sdkKey = configObj.sdkKey || '';
```

keppel2 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+      let cond = '';
+      conditions.forEach((item) => {
+        let subAudience = '';
+        // Checks if item is list of conditions means if it is sub audience
```
`// Checks if item is list of conditions means it is sub audience`

zashraf1985 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -13,15 +13,23 @@
+  OptimizelyAttribute,
+  OptimizelyAudience,
+  OptimizelyEvent,
+  OptimizelyExperiment,
+  OptimizelyVariable,
+  OptimizelyVariation,
+  Rollout,
   VariationVariable,
   Variation,
-  Rollout,
```
NIT: arrange these in alphabetical order

zashraf1985 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -37,20 +45,31 @@ export class OptimizelyConfig {
+  public audiences: OptimizelyAudience[];
+  public events: OptimizelyEvent[];
   private datafile: string;
+  private static audienceConditionalOperators: string[] = ['and', 'or', 'not'];
```
This can probably moved to ENUMS?

zashraf1985 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+   * @returns     {OptimizelyAudience[]}       Array of unique audiences
+   */
+  static getAudiences(configObj: ProjectConfig): OptimizelyAudience[] {
+    const finalAudiences: OptimizelyAudience[] = [];
```
why call it `finalAudiences`? can you simply call it `audiences`

zashraf1985 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+   */
+  static getAudiences(configObj: ProjectConfig): OptimizelyAudience[] {
+    const finalAudiences: OptimizelyAudience[] = [];
+    const typedAudienceIds: { [id: string]: boolean } = {};
```
instead of a key, val pair, you could keep it an array of string ids and could use `typedAudienceIds.include(id)` to check if an id exists.

zashraf1985 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+    });
+
+    (configObj.audiences || []).forEach((oldAudience) => {
+      if (!typedAudienceIds[oldAudience.id] && oldAudience.id != '$opt_dummy_audience') {
```
`oldAudience` can be simple named `audience` here.

zashraf1985 replied:
if you use array instead of object, you can simple do `!typedAudienceIds.includes(audience.id)` here.

zashraf1985 commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+      conditions.forEach((item) => {
+        let subAudience = '';
+        // Checks if item is list of conditions means if it is sub audience
+        if (item instanceof Array) {
+          subAudience = OptimizelyConfig.GetSerializedAudiences(item, audiencesById);
+          subAudience = `(${subAudience})`;
+        } else if (OptimizelyConfig.audienceConditionalOperators.includes(item)) {
+          cond = item.toUpperCase();
+        } else {
+          // Checks if item is audience id
+          const audienceName = audiencesById[item] ? audiencesById[item].name : item;
+          // if audience condition is "NOT" then add "NOT" at start. Otherwise check if there is already audience id in serializedAudience then append condition between serializedAudience and item
+          if (serializedAudience || cond === 'NOT') {
+            cond = cond === '' ? 'OR' : cond;
+            if (serializedAudience === '') {
+              serializedAudience = `${cond} "${audiencesById[item].name}"`;
+            } else {
+              serializedAudience = serializedAudience.concat(` ${cond} "${audienceName}"`);
+            }
+          } else {
+            serializedAudience = `"${audienceName}"`;
+          }
+        } 
+        // Checks if sub audience is empty or not
+        if (subAudience !== '') {
+          if (serializedAudience !== '' || cond === 'NOT') {
+            cond = cond === '' ? 'OR' : cond;
+            if (serializedAudience === '') {
+              serializedAudience = `${cond} ${subAudience}`;
+            } else {
+              serializedAudience = serializedAudience.concat(` ${cond} ${subAudience}`);
+            }
+          } else {
+            serializedAudience = serializedAudience.concat(subAudience);
+          }
+        }
+      });
+    }
+    return serializedAudience;
```
NOT reviewing this code. will leave it to @msohailhussain and @jaeopt as they have more context about this particular peice

yavorona commented on packages/optimizely-sdk/lib/shared_types.ts:
```diff
@@ -248,6 +249,7 @@ export interface OptimizelyOptions {
 export interface OptimizelyExperiment {
   id: string;
   key: string;
+  audiences: unknown[] | string;
```
Should just be `string`

yavorona commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+    configObj: ProjectConfig,
+    featureVariableIdMap: FeatureVariablesMap,
+    featureId: string,
+    experiments: Experiment[] | undefined
```
This should not be expected undefined as every rollout object should contain `experiments`. @msohailhussain please correct if I am wrong.

yavorona commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+    experiments: Experiment[] | undefined
+  ): OptimizelyExperiment[] {
+    const deliveryRules: OptimizelyExperiment[] = [];
+    if (!experiments) {
```
If above is true, this if statement is unnecessary.

yavorona commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -61,6 +80,250 @@ export class OptimizelyConfig {
+    isFeatureEnabled: boolean | undefined
+  ): OptimizelyVariablesMap {
+    let variablesMap: OptimizelyVariablesMap = {};
+    if (!featureId) {
```
This statement seems unnecessary as we should expect a valid `featureId`, right?

msohailhussain reviewed (changes requested):
please address

yavorona commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -62,157 +81,360 @@ export class OptimizelyConfig {
+    featureId: string
+  ): { [key: string]: Variation } {
+    let variationsMap: { [key: string]: OptimizelyVariation } = {};
+    variationsMap = variations.reduce((OptlyVariationsMap: { [key: string]: OptimizelyVariation }, variation) => {
```
`optlyVariationsMap` instead of `OptlyVariationsMap`

yavorona commented on packages/optimizely-sdk/lib/core/optimizely_config/index.ts:
```diff
@@ -62,157 +81,360 @@ export class OptimizelyConfig {
+    isFeatureEnabled: boolean | undefined
+  ): OptimizelyVariablesMap {
+    const variablesMap = (featureIdVariableMap[featureId] || []).reduce(
+      (OptlyVariablesMap: OptimizelyVariablesMap, featureVariable) => {
```
`optlyVariablesMap` instead of `OptlyVariablesMap`

yavorona reviewed (commented):
Looks good except for couple of minor comments and failing browser stack tests.

jaeopt commented on packages/optimizely-sdk/lib/core/project_config/index.ts:
```diff
@@ -81,7 +83,7 @@ export interface ProjectConfig {
   groups: Group[];
   events: Event[];
-  attributes: Array<{ id: string }>;
+  attributes: OptimizelyAttribute[];
```
I think it's not a good idea to use "OptimizelyAttribute" or other "OptimizelyXYZ" types in here. 
Even if they look identical for now, we should separate them as public types only.

yasirfolio3 replied:
makes sense.

yasirfolio3 replied:
makes sense, thanks.

jaeopt reviewed (changes requested):
Overall looks great! 
- One comment about using OptimizelyConfig types internally. 
- Also I do not see duplicate-key warning about experimentsMap in OptimizelyConfig and deprecation warning about experimentsMap in OptimizelyFeature.

yasirfolio3 commented:
> Overall looks great!
> 
> * One comment about using OptimizelyConfig types internally.
> * Also I do not see duplicate-key warning about experimentsMap in OptimizelyConfig and deprecation warning about experimentsMap in OptimizelyFeature.

Had already added for `OptimizelyFeature` here: https://github.com/optimizely/javascript-sdk/pull/698/files#diff-2f8a7f1ccf360ce99abd03c811c59f9488fa24db24f8578d808e813e75192acdR330 

Also added for OptimizelyConfig Interface here but forgot about the class which i have added now. Comment on the interface can be seen here: https://github.com/optimizely/javascript-sdk/pull/698/files#diff-2f8a7f1ccf360ce99abd03c811c59f9488fa24db24f8578d808e813e75192acdR347

jaeopt reviewed (approved):
LGTM
nit - add a period (.) at the end of "This experimentsMap is for experiments of legacy projects only" in multiple places.
